controller/src/main/java/io/pravega/controller/store/stream/ZKGarbageCollector.java

```diff
@@ -0,0 +1,241 @@
+    private final String leaderPath;
+    private final String batchPath;
+    private final AtomicReference<NodeCache> watch;
+    private final AtomicReference<LeaderLatch> leaderLatch;
+    // dedicated thread executor for GC which will not interfere with rest of the processing.
+    private final ScheduledExecutorService gcExecutor;
+    private final long periodInMillis;
+    @VisibleForTesting
+    @Getter(AccessLevel.PACKAGE)
+    private final CompletableFuture<Void> acquiredLeadership = new CompletableFuture<>();
+    
+    ZKGarbageCollector(String gcName, ZKStoreHelper zkStoreHelper, Supplier<CompletableFuture<Void>> gcProcessing, Duration gcPeriod) {
+        Preconditions.checkNotNull(zkStoreHelper);
+        Preconditions.checkNotNull(gcProcessing);
+        Preconditions.checkArgument(gcPeriod != null && !gcPeriod.isNegative());
+
+        this.currentBatch = new AtomicLong();
+        this.gcName = gcName;
+        this.leaderLatch = new AtomicReference<>();
+        this.leaderPath = String.format(LEADER_PATH, gcName);
+        this.watch = new AtomicReference<>();
+        this.batchPath = String.format(BATCH_PATH, gcName);
+        this.zkStoreHelper = zkStoreHelper;
+        this.gcProcessing = gcProcessing;
+        this.periodInMillis = gcPeriod.toMillis();
+        this.gcExecutor = Executors.newSingleThreadScheduledExecutor();
+        this.gcLoop = new AtomicReference<>();
+    }
+
+    @Override
+    protected void doStart() {
+        long quantized = computeNewBatch();
+        byte[] b = getBytes(quantized);
+        // 1. create znode for storing latest batch id. If the batch id exists, get the value from the store.
+        // We will later register watch on the path and keep receiving any changes to its value.
+        zkStoreHelper.createZNodeIfNotExist(batchPath, b)
+                .thenCompose(x -> zkStoreHelper.getData(batchPath))
+                .thenAccept(data -> {
+                    currentBatch.set(BitConverter.readLong(data.getData(), 0));
+                })
+                .thenRun(this::initialize)
+                .whenComplete((r, e) -> {
+                    if (e != null) {
+                        notifyFailed(e);
+                    } else {
+                        notifyStarted();
+                    }
+                    latch.complete(null);
+                });
+    }
+
+    @Override
+    protected void doStop() {
+        latch.thenAccept(v -> {
+            if (gcLoop.get() != null) {
+                gcLoop.get().cancel(true);
+                gcLoop.get().whenComplete((r, e) -> {
+                    if (e != null) {
+                        notifyFailed(e);
+                    } else {
+                        notifyStopped();
+                    }
+                });
+            } else {
+                notifyStopped();
+            }
+        });
+    }
+
+    public Long getLatestBatch() {
+        return currentBatch.get();
+    }
+
+    @SneakyThrows(Exception.class)
```

Comment:
My comment about SneakyThrows was meant to refer to the fact that it is not a general purpose tool for turning exceptions into runtime ones. It's a very sharp tool and should be used carefully. Per our guidelines: https://github.com/pravega/pravega/wiki/Contributing#catching-exceptions

> For the specific case where you know with certainty an exception cannot occur, (i.e., the IOException when writing to a ByteArrayOutputStream) @SneakyThrows can be used to force the compiler to treat the exception as a runtime exception when it is naturally a checked exception. Because it hides the exception from callers @SneakyThrows should be used judiciously, only where this behavior is desired.

I don't think that applies in this situation. Or if it does can you greatly narrow the scope by using around a small try block for just the one method and narrowing to just the exception in it's signature that is impossible.